numba/npyufunc/_internal.c

```diff
@@ -104,6 +105,548 @@ PyTypeObject PyUFuncCleaner_Type = {
+    getargspec = PyObject_GetAttrString(inspect, "getargspec");
+    if (!getargspec) goto _get_nin_cleanup;
+    argspec = PyObject_CallFunctionObjArgs(getargspec, py_func_obj, NULL);
+    if (!argspec) goto _get_nin_cleanup;
+    args = PyObject_GetAttrString(argspec, "args");
+    if (!args) goto _get_nin_cleanup;
+    result = PyList_Size(args);
+
+  _get_nin_cleanup:
+    Py_XDECREF(args);
+    Py_XDECREF(argspec);
+    Py_XDECREF(getargspec);
+    Py_XDECREF(inspect);
+    return result;
+}
+
+static int
+dufunc_init(PyDUFuncObject *self, PyObject *args, PyObject *kws)
+{
+    PyObject *dispatcher=NULL, *keepalive=NULL, *py_func_obj=NULL, *tmp;
+    PyUFuncObject *ufunc=NULL;
+    int identity=PyUFunc_None;
+    int nin=-1, nout=1;
+    char *name=NULL, *doc=NULL;
+
+    static char * kwlist[] = {"dispatcher", "identity", "keepalive", "nin",
+                              "nout", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kws, "O|iO!nn", kwlist,
+                                     &dispatcher, &identity,
+                                     &PyList_Type, &keepalive, &nin, &nout)) {
+        return -1;
+    }
+
+    py_func_obj = PyObject_GetAttrString(dispatcher, "py_func");
+    if (!py_func_obj) {
+        return -1;
+    }
+
+    if (nin < 0) {
+        nin = (int)_get_nin(py_func_obj);
+        if ((nin < 0) || (PyErr_Occurred())) {
+            Py_XDECREF(py_func_obj);
+            return -1;
+        }
+    }
+
+    /* Construct the UFunc. */
+    tmp = PyObject_GetAttrString(py_func_obj, "__name__");
+    if (tmp) {
+        name = PyString_AsString(tmp);
+    }
+    Py_XDECREF(tmp);
+    tmp = PyObject_GetAttrString(py_func_obj, "__doc__");
+    if (tmp && (tmp != Py_None)) {
+        doc = PyString_AsString(tmp);
+    }
+    Py_XDECREF(tmp);
+    tmp = NULL;
+    Py_XDECREF(py_func_obj);
+    py_func_obj = NULL;
+    if (!name) {
+        return -1;
+    }
+    ufunc = (PyUFuncObject *)PyUFunc_FromFuncAndData(NULL, NULL, NULL, 0,
+                                                     nin, nout, identity,
+                                                     name, doc, 0);
+    if (!ufunc) {
+        return -1;
+    }
+
+    /* Construct a keepalive list if none was given. */
+    if (!keepalive) {
+        keepalive = PyList_New(0);
+        if (!keepalive) {
+            Py_XDECREF(ufunc);
+            return -1;
+        }
+    } else {
+        Py_INCREF(keepalive);
+    }
+
+    tmp = self->dispatcher;
+    Py_INCREF(dispatcher);
+    self->dispatcher = dispatcher;
+    Py_XDECREF(tmp);
+
+    tmp = (PyObject*)self->ufunc;
+    self->ufunc = ufunc;
+    Py_XDECREF(tmp);
+
+    tmp = self->keepalive;
+    /* Already incref'ed, either by PyList_New(), or else clause, both above. */
+    self->keepalive = keepalive;
+    Py_XDECREF(tmp);
+
+    self->frozen = 0;
+
+    return 0;
+}
+
+static PyMemberDef dufunc_members[] = {
+    {"dispatcher", T_OBJECT_EX, offsetof(PyDUFuncObject, dispatcher), 0,
```

Comment:
Fine with me.  I will modify all attributes except ufunc.  IMO, a public ufunc member is part of the API, especially since we've made it behave like a ufunc, but don't directly expose the ufunc attributes.